[core] Deduplicate packages

merceyz opened the pull request:
- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui-org/material-ui/blob/master/CONTRIBUTING.md#submitting-a-pull-request).

- Added `yarn deduplicate` which automatically deduplicates packages using [yarn-deduplicate](https://www.npmjs.com/package/yarn-deduplicate) then runs `yarn install`
- Made the CI fail if there are duplicated packages

Output:
```sh
$ yarn upgrade @babel/core && node scripts/postinstall.js
yarn upgrade v1.17.3
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@1.2.9: The platform "win32" is incompatible with this module.
info "fsevents@1.2.9" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > @material-ui/pickers@3.1.1" has unmet peer dependency "@date-io/core@^1.3.6".
warning " > jss-rtl@0.2.3" has unmet peer dependency "jss@^9.0.0".
[4/4] Rebuilding all packages...
success Saved lockfile.
success Saved 9 new dependencies.
info Direct dependencies
└─ @babel/core@7.5.4
info All dependencies
├─ @babel/core@7.5.4
├─ @babel/helpers@7.5.4
├─ @babel/highlight@7.5.0
├─ ansi-styles@3.2.1
├─ chalk@2.4.2
├─ color-convert@1.9.3
├─ color-name@1.1.3
├─ js-tokens@4.0.0
└─ supports-color@5.5.0
Done in 16.74s.
22 duplicated package(s) found
  Package "@babel/core" wants ^7.1.6 and could get 7.5.4, but got 7.4.4
  Package "@babel/core" wants ^7.4.4 and could get 7.5.4, but got 7.4.4
  Package "@babel/generator" wants ^7.1.2 and could get 7.5.0, but got 7.4.4
  Package "@babel/generator" wants ^7.4.0 and could get 7.5.0, but got 7.4.4
  Package "@babel/helpers" wants ^7.1.2 and could get 7.5.4, but got 7.4.4
  Package "@babel/parser" wants ^7.0.0 and could get 7.5.0, but got 7.4.4
  Package "@babel/parser" wants ^7.1.2 and could get 7.5.0, but got 7.4.4
  Package "@babel/parser" wants ^7.1.6 and could get 7.5.0, but got 7.4.4
  Package "@babel/parser" wants ^7.4.3 and could get 7.5.0, but got 7.4.4
  Package "@babel/traverse" wants ^7.0.0 and could get 7.5.0, but got 7.4.4
  Package "@babel/traverse" wants ^7.1.0 and could get 7.5.0, but got 7.4.4
  Package "@babel/traverse" wants ^7.4.3 and could get 7.5.0, but got 7.4.4
  Package "@babel/types" wants ^7.1.2 and could get 7.5.0, but got 7.4.4
  Package "@babel/types" wants ^7.2.0 and could get 7.5.0, but got 7.4.4
  Package "@babel/types" wants ^7.3.0 and could get 7.5.0, but got 7.4.4
  Package "@babel/types" wants ^7.4.0 and could get 7.5.0, but got 7.4.4
  Package "ms" wants ^2.0.0 and could get 2.1.2, but got 2.1.1
  Package "resolve" wants ^1.10.0 and could get 1.11.1, but got 1.10.1
  Package "resolve" wants ^1.10.1 and could get 1.11.1, but got 1.10.1
  Package "resolve" wants ^1.4.0 and could get 1.11.1, but got 1.10.1
  Package "resolve" wants ^1.5.0 and could get 1.11.1, but got 1.10.1
  Package "resolve" wants ^1.8.1 and could get 1.11.1, but got 1.10.1
Deduplicating package(s)
yarn install v1.17.3
[1/4] Resolving packages...
[2/4] Fetching packages...
info fsevents@1.2.9: The platform "win32" is incompatible with this module.
info "fsevents@1.2.9" is an optional dependency and failed compatibility check. Excluding it from installation.
[3/4] Linking dependencies...
warning " > @material-ui/pickers@3.1.1" has unmet peer dependency "@date-io/core@^1.3.6".
warning " > jss-rtl@0.2.3" has unmet peer dependency "jss@^9.0.0".
[4/4] Building fresh packages...

$ node scripts/postinstall.js
Done in 12.57s.
```

eps1lon commented:
Can't lerna sync those?

Updating dependencies should probably be handled by a bot. Though it should only run weekly given the amount of packages in the workspace.

merceyz commented:
> Can't lerna sync those?

As far as I can tell, no.

> Updating dependencies should probably be handled by a bot. Though it should only run weekly given the amount of packages in the workspace.

Probably, but that isn't the goal of this PR

oliviertassinari commented on packages/material-ui-icons/package.json:
```diff
@@ -38,7 +38,7 @@
   },
   "peerDependencies": {
-    "@material-ui/core": "^4.1.2",
+    "@material-ui/core": "^4.2.0",
```
Can we losen the version down to v4? What about using the minimum version required for the packages to work? This will prevent duplications. Updating @material-ui/styles dependencies 
 while having it the update not required has created friction in the past (duplication of the singleton).

merceyz replied:
Since it's in `peerDependencies` it shouldn't cause any issues

oliviertassinari replied:
Wouldn't it be the same with a dependency? For instance a few lines above, we have:
```diff
  "dependencies": {
    "@babel/runtime": "^7.2.0",
-   "@material-ui/utils": "^4.0.1",
+   "@material-ui/utils": "^4.1.0",
```

We don't need to increase the minimal version of the utils. It works with 4.0.1. docs doesn't need any of the new features. Unless you have further insight on why the version should be increased. I think that we would be better off using the oldest possible versions. However, this is the theory, in practice, it's a different story.

When releasing new versions of the library, it's our responsibility to increase the minimal versions as they are needed by the different packages. So far, it's done manually, this opens the door for human mistakes. I haven't done any significant errors so far, but it's only a matter of time before Murphy hits us. I'm eager to put automatic checks and constraints in place to reduce the entropy of the possible cases it can go wrong. I have limited experience with Lerna, I only use it on [react-swipeable-views](https://github.com/oliviertassinari/react-swipeable-views). From what I have observed, Lerna automatically detects packages that need a version bump, and cross-reference the bump in all the packages that depend on it. So, If I understand your proposal correctly, the changes enforce the same behavior. Lerna is not the only option, I remember that necolas has a simple release script: https://github.com/necolas/react-native-web/blob/master/scripts/release/publish.js. 

So overall, it seems that we have to balance using the oldest possible version (which is critical for material-ui/styles) with reducing the scope of mistake when making releases.

merceyz replied:
> So, If I understand your proposal correctly, the changes enforce the same behavior

The idea was more of: if core depends on `@babel/runtime@^7.2.0` then the rest of the packages should match that version so that we don't get mismatches/duplicates.

But that's not really how it works so i'll revert this change and just keep the dedupe behaviour for now

oliviertassinari commented on scripts/postinstall.js:
```diff
@@ -0,0 +1,70 @@
+const { execSync, spawn } = require('child_process');
+const path = require('path');
+const fs = require('fs');
+const deduplicate = require('yarn-deduplicate');
```
👍

oliviertassinari commented on scripts/postinstall.js:
```diff
@@ -0,0 +1,70 @@
+
+process.chdir(path.resolve(__dirname, '..'));
+
+function syncWorkspace() {
```
What problem does it solve?

merceyz replied:
The idea was to make sure we specified the same version across the packages, i'm perfectly fine with removing it and just keeping the deduping. I don't mind at all

eps1lon commented on package.json:
```diff
@@ -2,6 +2,7 @@
   "version": "4.2.0",
   "private": true,
   "scripts": {
+    "postinstall": "node scripts/postinstall.js",
```
I would rather only use this as a CI check. An install should never mutate lockfiles (see npm issues for new contributors).

Why would we need to run this on every install?

merceyz replied:
>Why would we need to run this on every install?

The hope was to prevent cases like this https://github.com/mui-org/material-ui/pull/16521#discussion_r301615614 i'm fine either way

eps1lon replied:
Sure, I wasn't doubting the use case. Just that I don't want to cause any confusion when installing for the first time. I don't want to find out how this behaves across different versions of yarn installed. A CI task (maybe with a "how to fix" section in CONTRIBTUNG.md) should prevent those issues from being merged?

merceyz replied:
> I don't want to find out how this behaves across different versions of yarn installed

There shouldn't be any, because the installed version doesn't matter to the results from `yarn-deduplicate`.

> Just that I don't want to cause any confusion when installing for the first time

The way it's set up right now, nothing will happen on install. It will run the script, see no duplicated packages and exit. The only time it will do anything is after `yarn add`. Since `yarn upgrade ...` doesn't trigger `postinstall` you'll only see something on the CI or if you run the script manually.
Since the CI will fail if there are duplicated packages you'll never accidentally merge with duplicated packages, reinforcing
> nothing will happen on install

eps1lon replied:
> The way it's set up right now, nothing will happen on install. It will run the script, see no duplicated packages and exit.

What if it does that is my whole point.

> here shouldn't be any

This isn't sufficient in the current state of npm/yarn + lockfiles.

Just move it into a CI task. This is the intention of the task (find problems when integrating changes) so it should live where it's intention lies.

Unless there are actual reasons to put it into postinstall. So far I've only hear "this should be fine".

merceyz replied:
I totally agree.

merceyz commented:
@eps1lon I've added it into the CI task and moved it away from postinstall.

To deduplicate packages one can now use `yarn deduplicate`.

Example of a CI failure where there was duplicated packages: https://circleci.com/gh/mui-org/material-ui/106720?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

oliviertassinari commented on scripts/deduplicate.js:
```diff
@@ -0,0 +1,64 @@
+}
+
+async function run() {
+  if (process.env.NOYARNPOSTINSTALL) {
```
Do we still need it?

merceyz replied:
Ah, forgot about that one.
I'll remove it, cheers

merceyz commented:
Only benefit it provides locally is that it runs `yarn install` for you while on the CI it doesn't write to the disk and provides a more useful error message.

oliviertassinari commented:
Both approaches sound OK, we had this duplication problems surface twice in the past on dependencies upgrade PR, great to see it fixed systematically. @eps1lon, your call.
